source/symbols/TypeHelpers.h

```diff
@@ -0,0 +1,208 @@
+            buffer.append(static_cast<byte>(*it));
+        len *= CHAR_BIT;
+        auto ci = SVInt(len, span<const byte>(buffer), false);
+        auto msb = len - bit % CHAR_BIT - 1;
+        auto lsb = CHAR_BIT - 1 - (bit + width - 1) % CHAR_BIT;
+        bit += width;
+        ASSERT(bit <= str.length() * CHAR_BIT);
+        if (lsb > 0 || msb < len - 1)
+            return ci.slice(static_cast<int32_t>(msb), static_cast<int32_t>(lsb));
+        else
+            return ci;
+    }
+}
+
+/// Performs unpack operation on a bit stream
+static ConstantValue unpack(const Type& type, PackVector::const_iterator& iter, bitwidth_t& bit,
+                            bitwidth_t& dynamic) {
+    if (type.isIntegral()) {
+        auto width = type.getBitWidth();
+        SmallVectorSized<SVInt, 8> buffer;
+        while (width > 0) {
+            auto ci = slicePacked(iter, bit, width);
+            ASSERT(ci.getBitWidth() <= width);
+            width -= ci.getBitWidth();
+            if (!type.isFourState())
+                ci.flattenUnknowns();
+            buffer.emplace(ci);
+        }
+        auto cc = SVInt::concat(span<SVInt const>(buffer.begin(), buffer.end()));
+        cc.setSigned(type.isSigned());
+        return cc;
+    }
+    if (type.isString()) {
+        if (!dynamic)
+            return std::string();
+        auto width = dynamic;
+        ASSERT(width % CHAR_BIT == 0);
+        dynamic = 0;
+        SmallVectorSized<SVInt, 8> buffer;
+        while (width > 0) {
+            auto ci = slicePacked(iter, bit, width);
+            ASSERT(ci.getBitWidth() <= width);
+            width -= ci.getBitWidth();
+            ci.flattenUnknowns();
+            buffer.emplace(ci);
+        }
+        return ConstantValue(SVInt::concat(span<SVInt const>(buffer.begin(), buffer.end())))
+            .convertToStr();
+    }
+    if (type.isUnpackedArray()) {
+        const auto& ct = type.getCanonicalType();
+        SmallVectorSized<ConstantValue, 16> buffer;
+        if (ct.kind != SymbolKind::FixedSizeUnpackedArrayType) {
+            if (dynamic > 0) {
+                auto elemWidth = type.getArrayElementType()->bitstreamWidth();
+                if (!elemWidth)
+                    elemWidth = dynamic;
+                ASSERT(dynamic % elemWidth == 0);
+                for (auto i = dynamic / elemWidth; i > 0; i--)
+                    buffer.emplace(unpack(*type.getArrayElementType(), iter, bit, dynamic));
+                ASSERT(!dynamic || type.getArrayElementType()->isFixedSize());
+                dynamic = 0;
+            }
+        }
+        else {
+            const auto& ct1 = ct.as<FixedSizeUnpackedArrayType>();
+            const auto& elem = ct1.elementType;
+            for (auto width = ct1.range.width(); width > 0; width--)
+                buffer.emplace(unpack(elem, iter, bit, dynamic));
+        }
+        return constContainer(ct, span<const ConstantValue>(buffer));
```

Comment:
Are the explicit span constructors necessary everywhere? I would expect them to work implicitly by just passing the buffer.

Reply:
I tried removing them, but compile failed. In one case, I simplified it into span<buffer>, clang/gcc accepted, but msvc failed.